Font Library: Font Collection backend

matiasbenedetto opened the pull request:
## What?
Add extensibility capabilities to the Font Library.
Provides a way to provide collections of typographic fonts by code. 


## Why?
- Standardize the way to add font collections to the font library.
- Allows extenders code (plugins, for example) to register lists of fonts to let the user install the ones they want. Font foundries, for example, will be able to create a plugin to provide their fonts with just a few lines of code.

## Example:

This is all the code required to create a plugin providing a font collection.

```php
<?php
/*
Plugin Name: My Font Collection
Plugin URI: https://your-website.com/
Description: Add a font collection to your WordPress Font Library.
Version: 1.0
Author: Your Name
Author URI: https://your-website.com/
License: GPLv2 or later
Text Domain: my-font-collection
*/

$my_config = array (
    'id'             => 'my-font-collection',
    'name'           => 'My Font Collection',
    'description'    => 'Demo about how to a font collection to your WordPress Font Library.',
    'data_json_file' => path_join( __DIR__, 'my-font-collection-data.json' ),
);

if ( function_exists( 'wp_register_font_collection' ) ) {
    wp_register_font_collection ( $my_config );
}

?>
```

Download the plugin as zip ready to test:
[my-font-collection.zip](https://github.com/WordPress/gutenberg/files/12443217/my-font-collection.zip)


## How?
- Implementation of a `WP_Font_Collection` class.
- Implementation of the `wp_register_font_collection` filter.
- Implementation of the `/fonts/collections` and `fonts/collections/<id>` endpoints.

## To do
- Add more ways to provide lists of fonts apart from the JSON file. (maybe for a future PR).
- Add sanitization.
We would need this: https://github.com/WordPress/gutenberg/issues/52798


## Tracking issue:
Stage 1: https://github.com/WordPress/gutenberg/issues/52698
Stage 2: https://github.com/WordPress/gutenberg/issues/53307

matiasbenedetto committed:
Adding Font Collection class

matiasbenedetto committed:
php formatting and linting

matiasbenedetto committed:
Adding tests for collections routes

matiasbenedetto committed:
Adding tests for WP_Font_Collection constructor

matiasbenedetto committed:
adding tests for WP_Font_Collection get_data()

matiasbenedetto committed:
adding tests for WP_Font_Library register_font_collection()

spacedmonkey commented on lib/experimental/fonts/font-library/class-wp-font-library.php:
```diff
@@ -27,6 +27,63 @@ class WP_Font_Library {
+	 *
+	 * @since 6.4.0
+	 */
+	public static function register_filters() {
```
Just use $this here.  No need for this to a static method

hellofromtonya replied:
@matiasbenedetto this method is not needed. Instead, move its code to `font-library.php` file.  Rather than invoking this method which then hooks into the filter, directly hook the filter without the method.

Less code to maintain.

matiasbenedetto replied:
>  move its code to font-library.php file. Rather than invoking this method which then hooks into the filter, directly hook the filter without the method.

Nice idea. Done!

spacedmonkey commented on lib/experimental/fonts/font-library/class-wp-font-library.php:
```diff
@@ -27,6 +27,63 @@ class WP_Font_Library {
+	 * @param array  $config Font collection config options.
+	 * @return WP_Font_Collection|WP_Error A font collection is it was registered succesfully and a WP_Error otherwise.
+	 */
+	public static function register_font_collection( $id, $config ) {
```
No need for this to be static.

hellofromtonya replied:
>No need for this to be static.

I respectfully disagree. Let me explain why.

The `WP_Font_Library` is not an object. Rather its public facing methods are a collection of shared functionality.

Rather than separate instances of objects with separate collections, the design is intentional to share one font collection across all of Gutenberg and soon Core.

The design also simplifies usage. How so?

* No instantiation is needed.
* No singleton control is needed.
* Consumers do not need to first get the object instance to then register their font collection.

Thus, IMO the architecture fits Core well and is intentional.

IMO static is appropriate here.

matiasbenedetto replied:
I tried to make this avoiding `static` and it seems to require more code to do the same. in my opinion, in this case, it also makes the code harder to read  and undersatand.

spacedmonkey commented on lib/experimental/fonts/font-library/class-wp-font-library.php:
```diff
@@ -27,6 +27,63 @@ class WP_Font_Library {
+	 *
+	 * @var array
+	 */
+	public static $collections = array();
```
No need for this to be static.

hellofromtonya replied:
I respectfully disagree. Reasoning is shared here https://github.com/WordPress/gutenberg/pull/53816/files#r1300561853.

spacedmonkey commented:
If we must use static, then follow the pattern in wp_metadata_lazyloader that creates a instance to a static variable.

matiasbenedetto committed:
get font collection tests

matiasbenedetto commented:
> If we must use static, then follow the pattern in wp_metadata_lazyloader that creates a instance to a static variable.

@spacedmonkey Thanks for the review!
Could you please add more context about why is that pattern preferred over static methods and properties?

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-library.php:
```diff
@@ -27,6 +27,78 @@ class WP_Font_Library {
+	 */
+	public static function register_filters() {
+		add_filter(
+			'register_font_collection',
```
Prefix the filter's name with `wp_` to avoid naming collisions in the wild.

matiasbenedetto replied:
Done!

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-library.php:
```diff
@@ -27,6 +27,78 @@ class WP_Font_Library {
+	 *
+	 * @var array
+	 */
+	public static $collections = array();
```
I think this should be `private`, not `public`. There are public methods for interacting with this static property.

matiasbenedetto replied:
Done!

matiasbenedetto committed:
adding 'wp_' prefix to the 'register_font_collection' filter name

matiasbenedetto committed:
fix callback name

matiasbenedetto committed:
making class property private

matiasbenedetto committed:
registering filter from font-library.php file

matiasbenedetto committed:
Merge branch 'trunk' into try/font-collections

matiasbenedetto committed:
removing superfluous comment

matiasbenedetto committed:
moving files to according changes in trunk

matiasbenedetto committed:
config without a json file should fail

matiasbenedetto committed:
fix property name in tests

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-collection.php:
```diff
@@ -0,0 +1,90 @@
+ * This file contains the Font Collection class definition.
+ *
+ * @package    WordPress
+ * @subpackage Font Collection
```
```suggestion
 * @subpackage Font Library
```

It's part of the Font Library.

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-collection.php:
```diff
@@ -0,0 +1,90 @@
+	 *
+	 * @param string $id Font collection id.
+	 * @param array  $config Font collection config options.
+	 * @throws Exception If some of the required parameters are missing.
```
```suggestion
	 * @throws Exception If the required parameters are missing.
```

matiasbenedetto committed:
name fix

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

matiasbenedetto committed:
comment update

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-collection.php:
```diff
@@ -0,0 +1,90 @@
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return array An array contaning the list of font families in theme.json format
+	 */
+	public function get_data() {
+		if ( ! empty( $this->config['data_json_file'] ) ) {
+			if ( file_exists( $this->config['data_json_file'] ) ) {
+				$data                    = file_get_contents( $this->config['data_json_file'] );
+				$collection_data         = $this->get_config();
+				$collection_data['data'] = $data;
+				unset( $collection_data['data_json_file'] );
+				return $collection_data;
+			}
+		}
+		return new WP_Error( 'font_collection_error', 'Font Collection data is missing the data_json_file.' );
```
This can be improved to further guard against possible failure points such as:

* `file_get_contents()` can return `false` on failure. 
* `file_get_contents()` returns an empty string when there's nothing in the file.

~* What if the config array does not have a `'data'` key / element? This can be solved by setting up the initial `$config` in the constructor.~

@matiasbenedetto let me push a suggestion for your consideration, as it's likely easier to visualize what I'm suggesting by viewing the code  changes.

hellofromtonya replied:
https://github.com/WordPress/gutenberg/pull/53816/commits/85fe753e506bbf043b32d6cd14022fc83e0ecf21 reworks the `get_data()` method by:

* Removing the first `empty()` check -> already checked in the constructor.
* Moves the `WP_Error` act as an entry point guard clause. If the data JSON file does not exist, bail out and return the error immediately, as there's nothing more to do in the method.
* Adds error checking for `file_get_contents()`.

What do you think @matiasbenedetto ?

matiasbenedetto replied:
Looking good!

hellofromtonya committed:
Adds WP_Error to return type

hellofromtonya committed:
Improves contructor error handling.

* Ensures each required param is of the right data type.
* Improves each param check error to include expected data type.

hellofromtonya committed:
FontCollection::get_data(): Improves error handling

* Rechecking if "data_json_file" exists in the $config property shouldn't be necessary,
  as it's checked in the constructor.

* If the file does not exist, bail out immediately as there's nothing more to do.

* Adds a check and WP_Error for file_get_contents():
file_get_contents() returns false on failure.
If there's nothing in the file, an empty string is returned.
This change checks for both of these conditions and returns a WP_Error if either happens.

* Internationalizes WP_Error error message for consistency in Core.

hellofromtonya committed:
Removes empty space for wpcs

hellofromtonya commented on lib/experimental/fonts/font-library/font-library.php:
```diff
@@ -33,5 +33,20 @@ function gutenberg_init_font_library() {
-add_action( 'rest_api_init', 'gutenberg_init_font_library' );
+add_action( 'rest_api_init', 'gutenberg_init_font_library_routes' );
 
+/**
+ * Registers the font collection filter for font library.
+ *
+ * @since 6.4.0
+ */
+function gutenberg_add_register_font_collection_filter() {
+	add_filter(
+		'wp_register_font_collection',
+		array( 'WP_Font_Library', 'register_font_collection' ),
+		10,
+		2
+	);
+}
+
+add_action( 'init', 'gutenberg_add_register_font_collection_filter' );
```
Why does it need to add the filter on `'init'`? Why not move the `add_filter` hook out of the function like this?

```suggestion
add_filter( 'wp_register_font_collection', array( 'WP_Font_Library', 'register_font_collection' ), 10, 2 );
```

matiasbenedetto replied:
Good suggestion!

matiasbenedetto committed:
adding filter in a simpler way

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-library.php:
```diff
@@ -27,6 +27,64 @@ class WP_Font_Library {
+	 * @return WP_Font_Collection|WP_Error A font collection is it was registered succesfully and a WP_Error otherwise.
+	 */
+	public static function register_font_collection( $id, $config ) {
+		if ( array_key_exists( $id, self::$collections ) ) {
```
```suggestion
		if ( isset( self::$collections[ $id ] ) ) {
```

Tiny tiny tiny micro-optimization.

matiasbenedetto committed:
micro-optimization

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-collection.php:
```diff
@@ -0,0 +1,93 @@
+	 */
+	public function __construct( $id, $config ) {
+
+		if ( empty( $id ) && is_string( $id ) ) {
```
The `try/catch` eliminates the need to throw a fatal error here.

To eliminate the `try/catch, what if instead `WP_Error` is used and returned? That would eliminate the `try/catch` code.

matiasbenedetto replied:
Could you provide an example of what you mean, please?

hellofromtonya replied:
Absolutely!

hellofromtonya replied:
There are a couple of ways to do this:

* Return `WP_Error()`.
* Create an InvalidArgument exception that returns `WP_Error()`.

I think for now the first option is more straightforward.

See commit https://github.com/WordPress/gutenberg/pull/53816/commits/0e6c026c87933f64dae78049cc9bb53c7fa8dffb.

What do you think @matiasbenedetto ?

matiasbenedetto replied:
I thought about that before but after some little research, I think that returning values from a PHP class __construct method as in https://github.com/WordPress/gutenberg/pull/53816/commits/0e6c026c87933f64dae78049cc9bb53c7fa8dffb isn't the best approach. I made a very simple snippet to test the idea without any other context:

```
class MyError {
  private $type;

  public function __construct($type) {
    $this->type=$type;
  }

  public function print_error_type() {
    echo $this->type;
  }

}

class TestingClass{
  public function __construct () {
    return new MyError("ERROR MESSAGE");
  }

}

$error1 = new MyError("Hello!");
echo ( get_class( $error1 ) ); //MyError
$error1->print_error_type(); //Hello!


$error2 = new TestingClass();
echo ( get_class( $error2 ) ); //TestingClass
$error2->print_error_type(); //PHP Error:  Call to undefined method TestingClass::print_error_type()
```
I think this example shows that it won't work as expected because the class of the `$error2 = new TestingClass();` is still `TestingClass` no matter if it's returning a `MyError` object.

hellofromtonya replied:
Fair points. If the class is instantiated outside of the Font Library's usage, then yes, unexpected results can happen.

Let me see how this handled within Core for some inspiration. I'll revert or modify the approach.

hellofromtonya replied:
Wait duh 🤦‍♀️ Oh silly me ... of course. The constructor returns that instance of its class. Imagine me flush faced 🤣 

Let's back up. What is the purpose of the `try/catch`? 

It's avoiding a fatal error and then recasting to an instance of `WP_Error`. 

Is that necessary? How are other endpoints handling fatal errors? Hmm, I'm curious. Time to explore other endpoints in Core.

hellofromtonya replied:
Reverted my changes that removed the `try/catch` https://github.com/WordPress/gutenberg/pull/53816/commits/f79212d86d3fd483334b4759aa0d2b1dca22d47d.

hellofromtonya replied:
Another way to go .. thinking out loud ... is to very the `$id` and `$config` is correct _before_ instantiating `WP_Font_Collection`, i.e. check both in `WP_Font_Library::register_font_collection()`. By doing so, the `try/catch` gets removed and the validation happens where it's passed in for registration.

Let me do an experiment.

hellofromtonya replied:
Aha, inspiration ...

What's bothering me is why the `try/catch` is needed. In reviewing, it is not needed. Registering a font collection happens within the extender's plugin or theme as they will invoke `WP_Font_Library::register_font_collection()`. It does not go through REST. So a fatal error is appropriate and will provide the developer the information they need to debug their custom code.

In going further into the design, the current implementation uses a filter approach for extenders to register their font collections. This can be simplified by removing the filter and instead directly calling `WP_Font_Library::register_font_collection()`. Why? It'll be more performant (no hooks to call and process) while giving a more clear stack backtrace if the `$id` or `$config` arguments are invalid - with a fatal error thrown.

@matiasbenedetto what do you think?

hellofromtonya replied:
Thinking about this more ... the preferred approach in Core is for extenders to use a global function, rather than the static method call:

Rather than:
```
WP_Font_Library::register_font_collection( $id, $config );
```

instead extenders call:
```
wp_register_font_collection( $id, $config );
```

matiasbenedetto replied:
Cool, based on you latest comments, this commit https://github.com/WordPress/gutenberg/pull/53816/commits/4e952e3360a42c97b09cbd21c183827130abdfec removes the  `wp_register_font_collection` filter and implements the function with the same name. It also removes the try/catch from the `WP_Font_Library::register_collection()` method, so if the `$config` has something wrong, a fatal error will be shown.

matiasbenedetto replied:
Apart from that, the replacement of the filter by the `wp_register_font_collection()` function allow us to use just one parameter (`$config`) instead of the previous 2 (`$id, $config`). I added a little refactor around that too in https://github.com/WordPress/gutenberg/commit/4e952e3360a42c97b09cbd21c183827130abdfec

hellofromtonya replied:
> function allow us to use just one parameter ($config) instead of the previous 2 ($id, $config)

Oh nice! 👏

hellofromtonya replied:
LGTM 👍 Well done @matiasbenedetto 👏

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-rest-font-library-controller.php:
```diff
@@ -67,6 +67,69 @@ public function register_routes() {
+		$collection = WP_Font_Library::get_font_collection( $id );
+
+		if ( is_wp_error( $collection ) ) {
+			return new WP_Error(
+				'font_collection_not_found',
+				__( 'Font collection not found.', 'gutenberg' ),
+				array( 'status' => 404 )
+			);
```
If an instance of `WP_Error` is returned, why not return that instance instead of creating a new one?

The `array( 'status' => 404 )` can be added to `WP_Error` in  `WP_Font_Library::get_font_collection()`. And then it can be returned.

```suggestion
			return $collection;
```

matiasbenedetto committed:
reuse WP_Error response instead of creating a new one

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

hellofromtonya committed:
Eliminates try/catch

hellofromtonya committed:
Revert "Eliminates try/catch" commit

This reverts commit 0e6c026c87933f64dae78049cc9bb53c7fa8dffb.

matiasbenedetto committed:
Remove wp_register_font_collection and replace it by a global function, remove try catch and raise the error if needed

matiasbenedetto committed:
adding function comment and guard agains re-declaration

matiasbenedetto committed:
php format

matiasbenedetto committed:
fixing docblock coments

matiasbenedetto committed:
removing param comment

matiasbenedetto committed:
re-adding parameter comment removed by mistake

matiasbenedetto committed:
format php

matiasbenedetto committed:
updating WP_Font_Collection __construct tests

matiasbenedetto committed:
updating WP_Font_Collection get_data tests

hellofromtonya commented on lib/experimental/fonts/font-library/class-wp-font-collection.php:
```diff
@@ -0,0 +1,90 @@
+	 * @throws Exception If the required parameters are missing.
+	 */
+	public function __construct( $config ) {
+		if ( empty( $config ) ) {
```
```suggestion
		if ( empty( $config ) || ! is_array( $config ) ) {
```

matiasbenedetto replied:
I'm not sure about this change. $config should be an array, not a string, right?

matiasbenedetto replied:
Haha sorry! My bad, I misread `is_array` for `is_string` :sweat_smile: the change looks good

matiasbenedetto committed:
updating tests

matiasbenedetto committed:
php format

matiasbenedetto committed:
array check

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

hellofromtonya committed:
Documents config array structure

hellofromtonya commented:
The source code LTGM and is ready 👍

Next: the tests need updating for the changes and then a code review. And then this PR is ready for merge.

Almost over the finish line with this PR.

matiasbenedetto committed:
adding tests for /fonts/collections endpoint

matiasbenedetto committed:
adding /fonts/collections/<id> endpoint test

matiasbenedetto committed:
format

matiasbenedetto committed:
format

matiasbenedetto committed:
add test for missing collection

matiasbenedetto committed:
removing not needed variables

matiasbenedetto committed:
Add more tests and split the test for WP_REST_Font_Library_Controller

matiasbenedetto committed:
lint

matiasbenedetto committed:
try to create dir for tests in CI

matiasbenedetto committed:
Merge branch 'trunk' into try/font-collections

matiasbenedetto commented:
I updated the plugin demo code in the PR description and the plugin's zip file based on the latest PR changes.

matiasbenedetto commented:
I created a follow-up of this PR to load data from URLs. It depends on this one: https://github.com/WordPress/gutenberg/issues/53992

hellofromtonya committed:
Tests: Core prep

* Renamed files to remove `-test` suffix (no longer needed).
* Removed extra empty line before closing the class.
* Used Core's `assertWPError()`.

hellofromtonya reviewed (approved):
The source code LGTM 👍 

The tests are good enough. I'll follow-up with a test improvement PR.

This PR is ready (assuming the CI jobs are happy).

hellofromtonya committed:
Doh I changed the wrong assertion

matiasbenedetto commented:
Thanks, @hellofromtonya, for your work here!

matiasbenedetto commented:
I'm closing this PR because it's blocked by [the changes requested](https://github.com/WordPress/gutenberg/pull/53816#issuecomment-1685132908) by @spacedmonkey around the static class properties. [I asked for more details](https://github.com/WordPress/gutenberg/pull/53816#issuecomment-1686881378) about why we would need to avoid that pattern in this PR, and @hellofromtonya added [some input about why the usage of the static property are intentional](https://github.com/WordPress/gutenberg/pull/53816#discussion_r1300561853). Two weeks later, we didn't receive an answer, so I assume that the reasons presented are valid. This is blocking us from continuing the work on Font Library intended to be included in WordPress 6.4 release.

Because of that, I'm closing this PR and opening a new one. Of course, we are free to further discuss this topic on new issues, PRs, the make blog, or slack.
